### PR TITLE
[ALLUXIO-2447]Update fileInfoCommand to support get directory's info

### PR DESCRIPTION
**Pull request by gjhkael**
https://alluxio.atlassian.net/browse/ALLUXIO-2447

### CODE REVIEW

**Comment by gjhkael** [comment]
Jenkins, test this please

File: shell/src/main/java/alluxio/shell/command/FileInfoCommand.java [hunk 1]
```diff
@@ -41,14 +43,11 @@ public String getCommandName() {
...
-
-    if (status.isFolder()) {
-      throw new InvalidPathException(path + " is a directory path so does not have file blocks.");
-    }
-
+  /**
+   * @param status the URIStatus to be print
+   * @throws IOException if an I/O error occurs
+   */
+  public void printFileInfo(URIStatus status) throws IOException {
```
Comment (gpang) [inline shell/src/main/java/alluxio/shell/command/FileInfoCommand.java]: can this be `private`?

File: shell/src/main/java/alluxio/shell/command/FileInfoCommand.java [hunk 2]
```diff
@@ -58,12 +57,34 @@ protected void runCommand(AlluxioURI path, CommandLine cl) throws AlluxioExcepti
   }
 
   @Override
+  protected void runCommand(AlluxioURI path, CommandLine cl) throws AlluxioException, IOException {
+    URIStatus status = mFileSystem.getStatus(path);
+    if (status.isFolder()) {
+      List<URIStatus> childrens = mFileSystem.listStatus(path);
```
Comment (gpang) [inline shell/src/main/java/alluxio/shell/command/FileInfoCommand.java]: I feel like this does not have to list the children. It should only print out the details of the directory. Listing of the children can already be done by the shell.

**Comment by gpang** [comment]
@gjhkael Also, can you show a sample output of a the command performed on a directory? Thanks!

**Comment by jsimsa** [comment]
@gpang would it make sense to rename the command to `getStatus` as that's effectively what is going on internally?

File: shell/src/main/java/alluxio/shell/command/FileInfoCommand.java (hunk 1 above)
Comment (gjhkael) [inline shell/src/main/java/alluxio/shell/command/FileInfoCommand.java]: done

**Comment by gjhkael** [comment]
If the path is a directory, the output as following:
/tmp/1111 is a directory and have 2 child
The directory's information as following: 
FileInfo{fileId=9, name=1111, path=/tmp/1111, ufsPath=hdfs://svr4045hw2285.hadoop.lpt.qa.nt.ctripcorp.com:54310/tmp/1111, length=2, blockSizeBytes=0, creationTimeMs=1483581661717, completed=true, folder=true, pinned=false, cacheable=false, persisted=true, blockIds=[], inMemoryPercentage=0, lastModificationTimesMs=1483581663582, ttl=-1, ttlAction=DELETE, owner=op1, group=hadoop, mode=493, persistenceState=PERSISTED, mountPoint=false, fileBlockInfos=[]}
The childrens' information as following: 

child 1: 
FileInfo{fileId=536870911, name=_SUCCESS, path=/tmp/1111/_SUCCESS, ufsPath=hdfs://svr4045hw2285.hadoop.lpt.qa.nt.ctripcorp.com:54310/tmp/1111/_SUCCESS, length=0, blockSizeBytes=134217728, creationTimeMs=1483581663585, completed=true, folder=false, pinned=false, cacheable=true, persisted=true, blockIds=[], inMemoryPercentage=100, lastModificationTimesMs=1483581663586, ttl=-1, ttlAction=DELETE, owner=op1, group=hadoop, mode=420, persistenceState=PERSISTED, mountPoint=false, fileBlockInfos=[]}
Containing the following blocks: 

child 2: 
FileInfo{fileId=520093695, name=1.log, path=/tmp/1111/1.log, ufsPath=hdfs://svr4045hw2285.hadoop.lpt.qa.nt.ctripcorp.com:54310/tmp/1111/1.log, length=3328432, blockSizeBytes=134217728, creationTimeMs=1483581663578, completed=true, folder=false, pinned=false, cacheable=true, persisted=true, blockIds=[503316480], inMemoryPercentage=0, lastModificationTimesMs=1483581663578, ttl=-1, ttlAction=DELETE, owner=op1, group=hadoop, mode=420, persistenceState=PERSISTED, mountPoint=false, fileBlockInfos=[FileBlockInfo{blockInfo=BlockInfo{id=503316480, length=3328432, locations=[]}, offset=0, ufsLocations=[svr4045hw2285.hadoop.lpt.qa.nt.ctripcorp.com, svr4046hw2285.hadoop.lpt.qa.nt.ctripcorp.com]}]}
Containing the following blocks: 
BlockInfo{id=503316480, length=3328432, locations=[]}

If the path is a file, the output as following:
/tmp/1111/1.log is a file and the information as following: 
FileInfo{fileId=520093695, name=1.log, path=/tmp/1111/1.log, ufsPath=hdfs://svr4045hw2285.hadoop.lpt.qa.nt.ctripcorp.com:54310/tmp/1111/1.log, length=3328432, blockSizeBytes=134217728, creationTimeMs=1483581663578, completed=true, folder=false, pinned=false, cacheable=true, persisted=true, blockIds=[503316480], inMemoryPercentage=0, lastModificationTimesMs=1483581663578, ttl=-1, ttlAction=DELETE, owner=op1, group=hadoop, mode=420, persistenceState=PERSISTED, mountPoint=false, fileBlockInfos=[FileBlockInfo{blockInfo=BlockInfo{id=503316480, length=3328432, locations=[]}, offset=0, ufsLocations=[svr4045hw2285.hadoop.lpt.qa.nt.ctripcorp.com, svr4046hw2285.hadoop.lpt.qa.nt.ctripcorp.com]}]}
Containing the following blocks: 
BlockInfo{id=503316480, length=3328432, locations=[]}

File: shell/src/main/java/alluxio/shell/command/FileInfoCommand.java (hunk 2 above)
Comment (gjhkael) [inline shell/src/main/java/alluxio/shell/command/FileInfoCommand.java]: I do not find a command that can list all children of a directory. So, I think we can add a option for user to choice whether listing all the children of the path.

Comment (gpang) [inline shell/src/main/java/alluxio/shell/command/FileInfoCommand.java]: @gjhkael Can you just use `./bin/alluxio fs ls /path/to/dir/` to list the children of a directory?

**Comment by gpang** [comment]
@jsimsa @calvinjia Is it ok to add or modify the shell commands to `getStatus`? I think it can be confusing for `fileInfo` to return info for directories.

**Comment by jsimsa** [comment]
I would be personally be okay with deprecating `fileInfo` and introducing `getStatus` in its place.

**Comment by jsimsa** [comment]
Possibly calling it `stat` to continue the analogy with local file system.

File: shell/src/main/java/alluxio/shell/command/FileInfoCommand.java (hunk 2 above)
Comment (gjhkael) [inline shell/src/main/java/alluxio/shell/command/FileInfoCommand.java]: I know ./bin/alluxio fs ls /path/to/dir/ can list all children of a directory, but the detail info can not displays. So, I use a new options for user to decide whether need to show all the directory's children, and default just displays directory's info.

**Comment by gjhkael** [comment]
stat seems suitable.

**Comment by gpang** [comment]
@gjhkael I have a question about the behavior of the command. Is there a reason why `fileInfo` on a directory needs to print out its children? Would just printing info about the directory itself (and no children) be enough? Thanks!

**Comment by calvinjia** [comment]
@gpang @gjhkael I'm also in favor of deprecating the `fileInfo` call since the name is not really reflective of what we want it to do.

**Comment by gjhkael** [comment]
After @calvinjia 's detailed interpretation, It does make sense.

File: shell/src/main/java/alluxio/shell/command/StatCommand.java [hunk 3]
```diff
@@ -25,45 +24,48 @@
...
 
 /**
- * Displays the file's all blocks info.
+ * Displays the path's info.
+ * If path is a directory it displays the directory's info.
+ * If path is a file, it displays the file's all blocks info.
  */
 @ThreadSafe
-public final class FileInfoCommand extends WithWildCardPathCommand {
+public final class StatCommand extends WithWildCardPathCommand {
```
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/StatCommand.java]: We should not simply rename `fileInfo` to `stat`. Instead, we should keep the old `fileInfo` command around, and have it print information similar to what the `loadMetadata` command does informing the user that the command is deprecated and what command should be used instead.

Comment (gjhkael) [inline shell/src/main/java/alluxio/shell/command/StatCommand.java]: Thanks for your point out this deficiency, I will fix it.

**Comment by gjhkael** [comment]
alluxio-bot, check this please

**Comment by gjhkael** [comment]
Jenkins, test this please

**Comment by jsimsa** [comment]
Jenkins, test this please

File: docs/_includes/Command-Line-Interface/stat.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+```bash
+# Displays file's info
```
Comment (jsimsa) [inline docs/_includes/Command-Line-Interface/stat.md]: replace `info` with `status`

File: docs/_includes/Command-Line-Interface/stat.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+$ ./bin/alluxio fs stat /data/2015/logs-1.txt
+#
+# Displays directory's info
```
Comment (jsimsa) [inline docs/_includes/Command-Line-Interface/stat.md]: ditto

**Comment by gjhkael** [comment]
Why the InStreamOptionsTest can not passed?

File: tests/src/test/java/alluxio/shell/command/StatCommandTest.java [hunk 6]
```diff
@@ -37,15 +37,15 @@ public void fileInfoNotExist() throws IOException {
   public void fileInfoWildCard() throws IOException, AlluxioException {
```
Comment (gpang) [inline tests/src/test/java/alluxio/shell/command/StatCommandTest.java]: This should probably be named `statFileWildCard` or something similar.

Are there any tests you could write for stat'ing directories?

Also, does directories with with wildcard?

File: tests/src/test/java/alluxio/shell/command/StatCommandTest.java [hunk 7]
```diff
@@ -22,12 +22,12 @@
 import java.io.IOException;
 
 /**
- * Tests for fileInfo command.
+ * Tests for stat command.
  */
-public final class FileInfoCommandTest extends AbstractAlluxioShellTest {
+public final class StatCommandTest extends AbstractAlluxioShellTest {
   @Test
   public void fileInfoNotExist() throws IOException {
```
Comment (gpang) [inline tests/src/test/java/alluxio/shell/command/StatCommandTest.java]: This should probably be named `statFileNotExist` or something similar.

File: tests/src/test/java/alluxio/shell/command/StatCommandTest.java (hunk 6 above)
Comment (gjhkael) [inline tests/src/test/java/alluxio/shell/command/StatCommandTest.java]: done.
And directories with wildcard.
Also, I find a bug of stat or fileInfo command. see https://groups.google.com/forum/#!topic/alluxio-users/teg9KAh5rr8
I will create a new jira to fix it.

File: tests/src/test/java/alluxio/shell/command/StatCommandTest.java [hunk 8]
```diff
@@ -22,35 +22,46 @@
...
     Assert.assertTrue(res2.contains(testDir + "/foo/foobar2"));
     Assert.assertTrue(res2.contains(testDir + "/bar/foobar3"));
     Assert.assertFalse(res2.contains(testDir + "/foobar4"));
   }
+
+  @Test
+  public void statDirectoryWildCard() throws IOException, AlluxioException {
+    String testDir = AlluxioShellUtilsTest.resetFileHierarchy(mFileSystem);
+
+    mFsShell.run("stat", testDir + "/*");
```
Comment (gpang) [inline tests/src/test/java/alluxio/shell/command/StatCommandTest.java]: How is this test different from the first part of `statFileWildCard`? Do wildcards work the same regardless of files or directories?

File: docs/en/Command-Line-Interface.md [hunk 9]
```diff
@@ -184,15 +184,25 @@ which folders are taking up the most space.
...
+The `fileInfo` command dumps the FileInfo representation of a file to the console.
+It is primarily intended to assist powerusers in debugging their system. Generally viewing the file
+info in the UI will be much easier to understand.
 
 For example, `fileInfo` can be used to debug the block locations of a file. This is useful when
 trying to achieve locality for compute workloads.
 
 {% include Command-Line-Interface/fileInfo.md %}
 
+## stat
```
Comment (jsimsa) [inline docs/en/Command-Line-Interface.md]: keep the list of commands sorted alphabetically

Comment (gjhkael) [inline docs/en/Command-Line-Interface.md]: done

File: tests/src/test/java/alluxio/shell/command/StatCommandTest.java (hunk 8 above)
Comment (gjhkael) [inline tests/src/test/java/alluxio/shell/command/StatCommandTest.java]: I think the statFileWildCard is tests the files and directories, So the  statDirectoryWildCard test is necessary? Or let statDirectoryWildCard same with statFileWildCard?

**Comment by jsimsa** [comment]
LGTM

File: tests/src/test/java/alluxio/shell/command/StatCommandTest.java (hunk 8 above)
Comment (gjhkael) [inline tests/src/test/java/alluxio/shell/command/StatCommandTest.java]: @gpang What your opinion about it?

**Comment by gjhkael** [comment]
@calvinjia Can this PR merge to master?

File: docs/en/Command-Line-Interface.md [hunk 10]
```diff
@@ -184,9 +184,9 @@ which folders are taking up the most space.
 {% include Command-Line-Interface/du.md %}
 
 ## fileInfo
-The `fileInfo` command dumps the FileInfo representation of a file to the console. It is primarily
-intended to assist powerusers in debugging their system. Generally viewing the file info in the UI
-will be much easier to understand.
+The `fileInfo` command dumps the FileInfo representation of a file to the console.
+It is primarily intended to assist powerusers in debugging their system. Generally viewing the file
+info in the UI will be much easier to understand.
```
Comment (calvinjia) [inline docs/en/Command-Line-Interface.md]: Could you mention that this command is now deprecated in favor of `stat`?

File: shell/src/main/java/alluxio/shell/command/FileInfoCommand.java [hunk 11]
```diff
@@ -43,18 +40,9 @@ public String getCommandName() {
...
-    }
-
-    System.out.println(status);
-    System.out.println("Containing the following blocks: ");
-    AlluxioBlockStore blockStore = AlluxioBlockStore.create();
-    for (long blockId : status.getBlockIds()) {
-      System.out.println(blockStore.getInfo(blockId));
-    }
+    System.out
+        .println("The \"alluxio fs fileInfo <path>\" command is deprecated since version 1.4.");
```
Comment (calvinjia) [inline shell/src/main/java/alluxio/shell/command/FileInfoCommand.java]: `1.4`->`1.5`

Comment (calvinjia) [inline shell/src/main/java/alluxio/shell/command/FileInfoCommand.java]: Could you also mark this class as deprecated?

File: tests/src/test/java/alluxio/shell/command/StatCommandTest.java [hunk 12]
```diff
@@ -22,30 +22,51 @@
 import java.io.IOException;
 
 /**
- * Tests for fileInfo command.
+ * Tests for stat command.
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/shell/command/StatCommandTest.java]: Although `getFileInfo` is deprecated, I would prefer to keep the tests for it and make a new copy for testing `stat`.

**Comment by calvinjia** [comment]
@gjhkael I had a few more minor comments, otherwise this looks good.

**Comment by yupeng9** [comment]
@calvinjia has your comments been addressed?

**Pull request merged**